### ISSUE TITLE
interpolations on provider does not reflected in show -json command

### ISSUE THREAD

**Issue by eerkunt** [hashicorp/terraform#24059]
### Terraform Version
```
0.12.20
```

### Terraform Configuration Files
```hcl
provider "aws" {
  region = local.region
}
locals {
  region = "eu-west-1"
}
```

### Debug Output
Output of `terraform plan -out=plan.out && terraform show -json plan.out`
```
{
  "format_version": "0.1",
  "terraform_version": "0.12.20",
  "planned_values": {
    "root_module": { }
  },
  "configuration": {
    "provider_config": {
      "aws": {
        "name": "aws",
        "expressions": {
          "region": {
            "references": [
              "local.region"
            ]
          }
        }
      }
    },
    "root_module": { }
  }
}
```

### Expected Behavior
Interpolated `local.region` should be seen in the jsonified plan output. 

### Actual Behavior
`region` parameter of the provider referenced as `local.region` instead of the interpolated value, `eu-west-1`. There is also no way to see within the plan about the value of `local.region`

### Steps to Reproduce
```
terraform init && \
terraform plan -out=plan.out && \
terraform show -json plan.out
```

**Comment by jbardin** [comment]
Hi @eerkunt,

This is the expected output currently. Provider configurations are not stored in the state or plan, because they can contain sensitive data which may even be specific to individual terraform runs.

Can you expand in the use case here?

**Comment by eerkunt** [comment]
Hi @jbardin,

Agree dumping the whole `provider` object would be quite dangerous. The use case here is to have [an external tool](https://terraform-compliance.com) that parses plan output and can ensure the provider is restricted to specific `region`s, or maybe get the version of the provider for doing something else.

Can't really see any other use case other than these two.

**Comment by jbardin** [comment]
Thanks @eerkunt,

It seems the issue here is that locals are evaluated on demand, and the config is not present in the `show -json` output, tooling can't directly resolve the `local.region` reference.

**Comment by eerkunt** [comment]
Correct. We are parsing plan output and leaving terraform to resolve any variables or interpolations for not duplicating the same code/logic.

Any plan outputting non-sensitive information about the provider ? 

E.g.: region, alias, version, assume role arn, maybe few more

**Comment by eerkunt** [comment]
Same problem applies if we are going to use a `data` resource to provide a value to a provider, e.g. using a github provider, but retrieving authentication `token` via `aws` provider ;

```hcl
            "github": {
                "name": "github",
                "expressions": {
                    "organization": {
                        "references": [
                            "var.github_organisation"
                        ]
                    },
                    "token": {
                        "references": [
                            "data.aws_ssm_parameter.github_token"
                        ]
                    }
                }
            }
```

This means; if we have a tool parses plan output, in order to find the value of this parameter, we need to implement ALL possible intepolations that terraform currently have.

**Comment by egorshulga** [comment]
@jbardin Would it be possible to drop resolved local variables to the plan too?

(Of course, plan is an internal thing, but I've been developing a bit of tooling to help our project in migration to Terraform)

**Comment by jbardin** [comment]
@egorshulga, It's not likely to happen until we have a way to make use of the values, which would probably mean waiting for a new plan format internally which can capture more useful data from the planning process. We also have to be aware that locals can contain the same possibly sensitive values being used by providers, which were not exposed previously.

**Comment by blowfishpro** [comment]
Maybe this doesn't belong in the plan per se, but it would be useful for Terraform to have a way to output all of the resolved provider configuration in a machine-readable format.

**Comment by cveld** [comment]
@jbardin this is very related to my issue #35674 indeed! I tend to disagree with your remark that these locals are not outputted into the plan as all relevant info surely is in the plan. The plan can be executed without any terraform config. It only requires the provider binaries. We made a small python script to prove this. It digs in the binary plan format and figures out the values of the locals.

Why are you so keen on not outputting sensitive values? The binary plan is already deemed privileged, so therefore the json projection would be deemed the same.

Actually the outside variable values are already outputted in the json. I would say these mainly contain the privileged values such as access tokens coming from external systems.

Our use case for having all azurerm provider details is to be able to check if resources to be created already exist. And if so, generate an import block for them.

**Comment by jbardin** [comment]
@cveld 
> I tend to disagree with your remark that these locals are not outputted into the plan as all relevant info surely is in the plan

Locals were not deemed relevant to the plan, because the plan only records the changes that are to be made, basically only what is necessary to run the apply phase. The value of the local itself makes no functional difference in the result since it has no side effects. Since locals need to be reevaluated anyway because they could be unknown or ephemeral, they are not stored and always reevaluated if needed during apply. Locals and module inputs are essentially treated as temporary values used only as a means to get to our final result.

Provider configurations are similar, the config is ephemeral and the configuration used during plan is not an artifact to be used during apply. It needs to be reevaluated at apply because it might be different during apply, so sensitive/ephemeral values aside, a record of the provider config wouldn't be useful for the Terraform CLI.

If the plan artifact were extended to record extra information about data that was used only during plan, it would be a net new feature. The prior use cases for this type of data was mostly for policy enforcement, but presenting a config which _was_ used already, and not the config which will be used in the future wasn't useful at the time.

**Comment by cveld** [comment]
@jbardin 
When you inspect the binary plan file you just see the one local sitting there that is required to configure the azurerm provider's subscription id for the plan to run.
Would be great to see this one local also in the json output.
For now we will use the binary for our workflow in order to derive the Azure subscription id.

**Comment by jbardin** [comment]
@cveld, the plan does not store any evaluation results from locals (and if it did it would probably be a bug, since locals can be ephemeral). I'm guessing you're looking at the bundled configuration, and that local's expression is just the string literal from your example. If that's the case, that value is no different than looking directly at the configuration files yourself -- it's just the unevaluated bytes from the configuration files. If the local were to also be a non-literal value referencing another object (e.g. `region = var.default_region`), you would still have the same problem, only one level deeper.

**Comment by cveld** [comment]
@jbardin
My point is: So far we were assuming we only required the json file for our
internal tooling. But now we know the provider config is not fully
recoverable from the json file, and only with the binary plan file, we do
include the binary plan file as well. With the binary plan we are able to
resolve the provider config through the included local references. The
binary plan contains some kind of smallest version of the required locals.
Actually now I am thinking of it, the binary plan file could be slimmed
down even more to only contain var references.
